### ISSUE TITLE
UNT0014 generics problem with interfaces

### ISSUE THREAD

**Issue by bdovaz** [microsoft/Microsoft.Unity.Analyzers#349]
#### Bug description

In the [documentation of this rule](https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0014.md) it clearly states that if you pass an interface it will not fail and it is so, but if you do it with generics it will fail.

- Version of analyzers assembly: 1.20.0
- Analyzer rule: UNT0014
- Error (exception message, type, and callstack where applicable): 'IMyInterface' is not a Unity Component

#### To Reproduce

```csharp
Test<IMyInterface>();

void Test<T>() where T : class
{
  gameObject.GetComponents<T>(); // Issues UNT0014
}
```

#### Expected behavior

It should not report a problem.

**Comment by sailro** [comment]
cc @jbevain 

You are right. But this is the typical case of generics where we cannot use the meta-information provided directly for `GetComponent<T>` and where we should infer effective type based on the usage: `Test<IMyInterface>`. This is not what we are doing know (for simplicity and performance).

So, for consistency, I refactored the diagnostic with the following rule:
- when we know for sure that the usage is invalid, add a diagnostic
- when we know for sure that the usage is valid, do nothing
- when we need to infer on usages (like in this case), do nothing (for now)
